contracts/hackatom/src/contract.rs

```diff
@@ -122,12 +126,14 @@ pub fn init(
 }
 
 pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, HackError> {
+    deps.api.addr_validate(&msg.verifier)?;
```

Comment:
This is easier to forget than forcing to change format. However, it would also fit inside a `HandleMsg.validate` method easily, which is a pattern we can start enforcing/suggesting and which makes more sense when no transforms are involved (one pass to verify, second pass to transform is expensive, so we kind of "validated" most messages inline when transforming them)

Reply:
One thing I was thinking about that makes it harder to forget the validation is having an `Addr` and an `ValidatedAddr`, which contains the same data and types are mostly compatible. But `ValidatedAddr` can only be created from trusted sources, like `info.sender` and `deps.api.addr_validate`

Reply:
Interesting idea.

Reply:
In order to avoid introducing more types than necessary, you can also use `String` in request messages and let `deps.api.addr_validate` convert a string to `StdResult<Address>`.

Reply:
So, `admin: String` in messages and `admin: Addr` in state (and env, info). And we just compare Addr with Addr (and refuse to compare with String).

This seems quite clear actually

Reply:
Yeah, something like that.

> And we just compare Addr with Addr (and refuse to compare with String).

This would require you to create many many unchecked String -> Addr constructions in tests at least. I think it is fine to compare Addr and String, but require a `Addr` in places where it matters.

Reply:
> This would require you to create many many unchecked String -> Addr constructions in tests at least

I do this in lots of places. `HumanAddr::from("somebody")` or `somebody.into()` depending on context. Or compare `assert_eq!(result.as_str(), "somebody")'`.

We need to convert from `&str` literals to `String` anyway, so using an Addr constructor should be fine. Maybe not use from here, but rather `Addr::trusted("somebody")` or `Addr::validated("somebody")`